servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp

```diff
@@ -1451,10 +1451,15 @@ void RendererCanvasRenderRD::canvas_render_items(RID p_to_render_target, Item *p
 		normal_transform.columns[2] = Vector2();
 		_update_transform_2d_to_mat4(normal_transform, state_buffer.canvas_normal_transform);
 
-		state_buffer.canvas_modulate[0] = p_modulate.r;
-		state_buffer.canvas_modulate[1] = p_modulate.g;
-		state_buffer.canvas_modulate[2] = p_modulate.b;
-		state_buffer.canvas_modulate[3] = p_modulate.a;
+		bool use_linear_colors = texture_storage->render_target_is_using_hdr(p_to_render_target) && GLOBAL_GET("rendering/viewport/hdr_2d");
```

Comment:
The project setting should likely be checked in the constructor only once, and saved in a member variable, for performance reasons. Reading project settings for each render command is going to be expensive.

That being said I'm not sure this is the correct check. I haven't looked in depth but I believe `rendering/viewport/hdr_2d` only sets the default value for the root viewport. It can still be changed later on for any viewport with `RenderingServer::viewport_set_use_hdr_2d` and `Viewport::set_use_hdr_2d`.

I'm no rendering expert so TIWAGOS[^1], but it sounds to me like this might need to check the value for the target viewport. No idea how that's done at this stage of the render pipeline, @clayjohn might know.

[^1]: Take It With A Grain Of Salt

Reply:
I suspect the original fix was correct. Taking a look now. 

It seems like the demo project is treating the linear HDR Viewport's texture as if it is sRGB color texture. If so, that would be the source of the problem.

Reply:
Ya, the problem comes from the demo project. When HDR_2D is disabled on the root viewport, the root viewport expects to receive sRGB textures (it doesn't do any color conversion). However, the demo passes a linear HD texture which ends up looking incorrect. Simply doing the linear->sRGB conversion in the shader makes the output correct. 

This issue highlights a broader issue which is that mixing HDR_2D and non HDR content is still confusing for users and prone to error. So there is likely stuff we can do to improve the situation. 

As for this PR, we should not be checking the project setting, the code should be reverted back to the way it was

Reply:
Thank you! I have reverted the changes.

I just started contributing to Godot and I'm trying to fix good first issues by tinkering with the code through trial and error. I'm wondering if there are any resources to learn more about the rendering architecture of Godot, so I can avoid situation like this next time.